pkg/monitors/elasticsearch/query/datapoints.go

```diff
@@ -0,0 +1,286 @@
+package query
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/signalfx/golib/v3/datapoint"
+	"github.com/signalfx/signalfx-agent/pkg/utils"
+	log "github.com/sirupsen/logrus"
+)
+
+// Struct keeps track of info required at a level of recursion
+// An instance of this struct can be thought of as a datapoint
+// collector for a particular aggregation
+type dpCollector struct {
+	aggName       string
+	aggRes        aggregationResponse
+	aggsMeta      map[string]*AggregationMeta
+	sfxDimensions map[string]string
+}
+
+// Returns aggregation type
+func (dpC *dpCollector) getType() string {
+	return dpC.aggsMeta[dpC.aggName].Type
+}
+
+// Walks through the response, collecting dimensions and datapoints depending on the
+// type of aggregation at each recursive level
+func collectDatapoints(resBody HTTPResponse, aggsMeta map[string]*AggregationMeta, sfxDimensions map[string]string) []*datapoint.Datapoint {
+	out := make([]*datapoint.Datapoint, 0)
+	aggsResult := resBody.Aggregations
+
+	for k, v := range aggsResult {
+		// each aggregation at the highest level starts with an empty set of dimensions
+		out = append(out, (&dpCollector{
+			aggName:       k,
+			aggRes:        *v,
+			aggsMeta:      aggsMeta,
+			sfxDimensions: sfxDimensions,
+		}).recursivelyCollectDatapoints()...)
+	}
+
+	return out
+}
+
+func (dpC *dpCollector) recursivelyCollectDatapoints() []*datapoint.Datapoint {
+	sfxDatapoints := make([]*datapoint.Datapoint, 0)
+
+	// The absence of "doc_count" and "buckets" field is a good indicator that
+	// the aggregation is a metric aggregation
+	if isMetricAggregation(&dpC.aggRes) {
+		return dpC.collectDatapointsFromMetricAggregation()
+	}
+
+	// Recursively collect all datapoints from buckets at this level
+	for _, b := range dpC.aggRes.Buckets {
+		key, ok := b.Key.(string)
+
+		if !ok {
+			log.WithFields(
+				log.Fields{
```

Comment:
Set a logger on `dpCollector` that has a default field of `monitorType: elasticsearch-query`.  See other monitors for an example.

Reply:
This makes it easier to identify which monitor is emitting the warning if you have a lot configured.